imgix.go

```diff
@@ -74,22 +80,46 @@ func (c *Client) Host(path string) string {
 	return RegexpRemoveHTTPAndS.ReplaceAllString(host, "")
 }
 
-func (c *Client) URL(imgPath string) url.URL {
-	return url.URL{
-		Scheme: c.Scheme(),
-		Host:   c.Host(imgPath),
-		Path:   imgPath,
+func (c *Client) SignatureForPath(path string) string {
+	return c.SignatureForPathAndParams(path, url.Values{})
+}
+
+func (c *Client) SignatureForPathAndParams(path string, params url.Values) string {
+	if c.token == "" {
+		return ""
+	}
+
+	hasher := md5.New()
+	hasher.Write([]byte(c.token + path))
+
+	// Do not mix in the parameters into the signature hash if no parameters
+	// have been given
+	if len(params) != 0 {
+		hasher.Write([]byte("?" + params.Encode()))
 	}
+
+	return "s=" + hex.EncodeToString(hasher.Sum(nil))
 }
 
 func (c *Client) Path(imgPath string) string {
-	url := c.URL(imgPath)
-	return url.String()
+	return c.PathWithParams(imgPath, url.Values{})
 }
 
 func (c *Client) PathWithParams(imgPath string, params url.Values) string {
-	u := c.URL(imgPath)
-	u.RawQuery = pluckSupportedParams(params).Encode()
+	u := url.URL{
+		Scheme:   c.Scheme(),
+		Host:     c.Host(imgPath),
+		Path:     imgPath,
+		RawQuery: params.Encode(),
+	}
+
+	signature := c.SignatureForPathAndParams(imgPath, params)
+	if signature != "" && len(params) > 0 {
+		u.RawQuery = u.RawQuery + "&" + signature
+	} else if signature != "" && len(params) == 0 {
+		u.RawQuery = signature
+	}
```

Comment:
Signature will be `""` when a token is not given to the client. It's possible to use unsigned URLs, although we don't recommend it.

It's also important that the signature always be the **last** parameter when present, hence the gross string concatenation. I'll leave a comment about that.